### DISCUSSION TITLE
Deleting by `Query` has a different behavior compared to deleting an ORM instance when using many-to-many mapping

### DISCUSSION THREAD

**Discussion by cainmagi** [sqlalchemy/sqlalchemy#11662] [Usage Questions] [answered]
I learned `sqlalchemy` by myself before. I am a little bit frustrated on the many-to-many mapping. In my case, if an ORM model `Table` has a many-to-many mapping and I delete it, the records in the association table should be automatically removed, like this:
```python
# Table has a many-to-many member Table.table_another = [...]
table_query = session.query(Table).filter(...).all()
for table in table_query:
    session.delete(table)
session.commit()
```
The above codes work as expected. Both the `table` and the association records are removed. However, if I delete the tables by using the query directly,
```python
table_query = session.query(Table).filter(...).all()
table_query.delete()
session.commit()
```
Then, only the items chosen by `table_query` will be deleted. The records in the association table are not deleted.

To my understanding, I think deleting items by `Query` should have the same behavior if I delete the objects by iterating the query. However, this experiment show different behaviors. I am wondering which part I did wrong. The document does not tell me what I should do in this case. The following document **does not** apply to my task:
https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-foreign-key-on-delete-with-many-to-many-relationships

## My `sqlalchemy` version

* `sqlalchemy`: `2.0.31`
* `Python`: `3.10.13`
* `OS`: `Debian GNU/Linux 12 (bookworm)`

## Full codes
```python
from typing import List
from sqlalchemy import ForeignKey
from sqlalchemy import String
from sqlalchemy import Table, Column
from sqlalchemy.orm import DeclarativeBase, MappedAsDataclass
from sqlalchemy.orm import Mapped
from sqlalchemy.orm import mapped_column
from sqlalchemy.orm import relationship


class Base(DeclarativeBase, MappedAsDataclass):
    pass


association_table = Table(
    "association_table",
    Base.metadata,
    Column("tablea_id", ForeignKey("tablea.id"), primary_key=True),
    Column("tableb_id", ForeignKey("tableb.id"), primary_key=True),
)


class TableA(Base):
    __tablename__ = "tablea"
    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str] = mapped_column(String(30))
    tableb: Mapped[List["TableB"]] = relationship(
        secondary=association_table,
        back_populates="tablea",
        default_factory=list,
    )

    def __repr__(self) -> str:
        return f"TableA(id={self.id!r}, name={self.name!r})"


class TableB(Base):
    __tablename__ = "tableb"
    id: Mapped[int] = mapped_column(primary_key=True, init=False)
    name: Mapped[str] = mapped_column(String(30))
    tablea: Mapped[List["TableA"]] = relationship(
        secondary=association_table,
        back_populates="tableb",
        default_factory=list,
    )

    def __repr__(self) -> str:
        return f"TableB(id={self.id!r}, name={self.name!r})"


if __name__ == "__main__":
    import logging

    logging.getLogger("sqlalchemy.engine.Engine").disabled = True

    from sqlalchemy import create_engine
    from sqlalchemy import select
    from sqlalchemy.orm import Session

    engine = create_engine("sqlite://", echo=True)
    Base.metadata.create_all(engine)

    with Session(engine) as session:
        # Add TableA(vala1), TableA(valA2).
        ta_v1 = TableA(name="vala1")
        ta_v2 = TableA(name="vala2")
        session.add(ta_v1)
        session.add(ta_v2)
        session.commit()

        # Add TableB(valb1), TableB(valb2) to TableA(vala1).
        tb_v1 = TableB(name="valb1")
        tb_v2 = TableB(name="valb2")
        ta_v1 = session.query(TableA).filter(TableA.name == "vala1").first()
        if ta_v1 is not None:
            ta_v1.tableb.extend([tb_v1, tb_v2])
        session.add(tb_v1)
        session.add(tb_v2)
        session.commit()

        # Check all TableA, and the association values.
        ta_all = session.query(TableA).all()
        print(ta_all)  # [TableA(id=1, name='vala1'), TableA(id=2, name='vala2')]
        stmt = select("*").select_from(association_table)
        result = session.execute(stmt).fetchall()
        print(result)  # [(1, 1), (1, 2)]

        # Delete TableA(vala1).
        ta_v1_q = session.query(TableA).filter(TableA.name == "vala1")

        # If using this version, in the following check,
        # `result` will be `[(1, 1), (1, 2)]` while `ta_all` is `[TableA(id=2, name='vala2')]`
        ta_v1_q.delete()

        # If using this version, in the following check,
        # `result` will be `[]` while `ta_all` is `[TableA(id=2, name='vala2')]`
        # for ta_v1 in ta_v1_q.all():
        #     session.delete(ta_v1)
        session.commit()

        # Check all TableA, and the association values.
        ta_all = session.query(TableA).all()
        print(ta_all)
        stmt = select("*").select_from(association_table)
        result = session.execute(stmt).fetchall()
        print(result)
```

Please take a look at the line 91 and line 95, respectively. Here I use two different methods to delete the table, the results will differ according to how the `TableA(vala1)` is deleted.

**Comment by gordthompson** [comment]
The short answer is that without server-side FK `ON DELETE CASCADE` rules (a.k.a "passive deletes") cascades can happen automatically when an object is deleted from within the unit-of-work paradigm, i.e., `session.delete(ta_v1)`, but not when the row corresponding to the object is deleted directly. Hence warnings in the docs as in [this section](https://docs.sqlalchemy.org/en/20/orm/cascades.html#delete)

> Note that the ORM’s “delete” and “delete-orphan” behavior applies **only** to the use of the `Session.delete()` method to mark individual ORM instances for deletion within the unit of work process.

What does surprise me a bit is that the deletes are cascading without a `cascade=` attribute on the relationships. The [docs](https://docs.sqlalchemy.org/en/20/orm/cascades.html#cascades) say that "The default value of `relationship.cascade` is `save-update, merge.`" — i.e., no `delete` — so I would have expected SQLA to try setting the `association_table.tablea_id` values to NULL.

**Comment by cainmagi** [comment]
@gordthompson Thank you for explaining these to me. However, I still do not fully understand my issue. May I ask which `relationship` you are talking about when referring to `relationship.cascade`? Do you mean adding `ondelete=cascade` to `ForeignKey("tablea.id")` of `association_table`, or adding such configurations to `tableb: Mapped[List["TableB"]] = relationship(...)`? I have tried to follow this guide before:
https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-foreign-key-on-delete-with-many-to-many-relationships

However, it is not what I expected. Suppose that I add `cascade` to the relationship defined in `TableA`; it will automatically delete all related `TableB` rows when I delete a `TableA` item. However, what I am looking for is merely automatically deleting the rows in `association_table`. I want to preserve the `TableB` rows even if the only associated `TableA` is deleted.

As you have mentioned, even if I did not add any options about `cascade=delete` or `ondelete=cascade`, the rows in `association_table` can still be removed if I delete the `TableA` items one by one. In this process, does the automatic deleting of `association_table` rows work in a similar or even the same mechanism as cascade “delete” and “delete-orphan”? Please let me know. Thank you!

**Reply by gordthompson** [reply]
My apologies for confusing the issue re: cascades.

"what I am looking for is merely automatically deleting the rows in `association_table`. I want to preserve the `TableB` rows even if the only associated `TableA` is deleted."

That automatically happens for `session.delete(an_orm_object)` when the object has relationships that use an association table, as indicated by `secondary=`. As explained in the [docs](https://docs.sqlalchemy.org/en/20/orm/basic_relationships.html#deleting-rows-from-the-many-to-many-table):

> A behavior which is unique to the [relationship.secondary](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship.params.secondary) argument to [relationship()](https://docs.sqlalchemy.org/en/20/orm/relationship_api.html#sqlalchemy.orm.relationship) is that the [Table](https://docs.sqlalchemy.org/en/20/core/metadata.html#sqlalchemy.schema.Table) which is specified here is automatically subject to INSERT and DELETE statements, as objects are added or removed from the collection. There is no need to delete from this table manually.

I had forgotten that. In this special case, `cascade="all, delete"` on the relationship will refer to the subject of that relationship, e.g., "TableA" or "TableB", not the association table. Details [here](https://docs.sqlalchemy.org/en/20/orm/cascades.html#using-delete-cascade-with-many-to-many-relationships).

Note also that `ForeignKey(…, ondelete="CASCADE")` only affects the DDL for that column, e.g., when `Base.metadata.create_all(engine)` is used to create the table by emitting a CREATE TABLE statement.

**Reply by cainmagi** [reply]
Thank you! That's quite helpful. Now I have only two questions.

1. Is it designed as intended that the automatic deleting for the secondary table only happens when using `session.delete(an_orm_object)` but not happens when using `query.delete()`? I have checked the document you recommended. It seems that the document does not mention the expected behavior when using query to delete the table rows.
2. Suppose that I have used `query` to delete a table row by mistake. In this case, the association table will have some junks. Is it possible to prune the association table so that the "expired" records can be detected or removed. (Here I use "expired" to refer the records related to table rows that has been deleted).

**Reply by gordthompson** [reply, marked as answer]
>1. Is it designed as intended that the automatic deleting for the secondary table only happens when using `session.delete(an_orm_object)` but not happens when using `query.delete()`?

Yes, that is the expected behaviour.

>2. Suppose that I have used `query` to delete a table row by mistake. In this case, the association table will have some junks. Is it possible to prune the association table so that the "expired" records can be detected or removed. (Here I use "expired" to refer the records related to table rows that has been deleted).

```python
        # query to find orphaned rows in association_table
        find_orphans = (
            select(association_table)
            .outerjoin(TableA)
            .outerjoin(TableB)
            .where(or_(TableA.id.is_(None), TableB.id.is_(None)))
        )
        rows = session.execute(find_orphans).all()
        """
SELECT association_table.tablea_id, association_table.tableb_id 
FROM 
    association_table 
    LEFT OUTER JOIN tablea ON tablea.id = association_table.tablea_id 
    LEFT OUTER JOIN tableb ON tableb.id = association_table.tableb_id 
WHERE tablea.id IS NULL OR tableb.id IS NULL
        """
        print(rows)
        # [(1, 1), (1, 2)]

        # delete orphaned rows
        exists_a = (
            select(TableA.id)
            .where(TableA.id == association_table.c.tablea_id)
            .exists()
        )
        delete_a_orphans = delete(association_table).where(~exists_a)
        session.execute(delete_a_orphans)
        """
DELETE FROM association_table WHERE NOT (EXISTS (SELECT tablea.id 
FROM tablea 
WHERE tablea.id = association_table.tablea_id))
        """
        exists_b = (
            select(TableB.id)
            .where(TableB.id == association_table.c.tableb_id)
            .exists()
        )
        delete_b_orphans = delete(association_table).where(~exists_b)
        session.execute(delete_b_orphans)
        """
DELETE FROM association_table WHERE NOT (EXISTS (SELECT tableb.id 
FROM tableb 
WHERE tableb.id = association_table.tableb_id))
        """
        session.commit()
```

**Reply by cainmagi** [reply]
Thank you so much! I have totally solved my issue.

**Reply by gordthompson** [reply]
@cainmagi - I have updated the answer above to correct the "delete orphans" code.
